app/master/cluster_master.py

```diff
@@ -143,6 +144,15 @@ def handle_slave_state_update(self, slave, new_slave_state):
         do_transition = slave_transition_functions.get(new_slave_state)
         do_transition(slave)
 
+    def _shutdown_slave(self, slave):
+        """
+        Puts slave in shutdown mode so it cannot receive new builds. The slave will be killed when finished with any
+        running builds.
+        :type slave: Slave
+        """
+        slave.set_shutdown_mode()
+        self._logger.info('Slave on {} was put in shutdown mode. (id: {})', slave.url, slave.id)
```

Comment:
I like the idea of removing it from _all_slaves_by_url, but I'm not sure where to do the removal.  If we do it here, the slave can still be executing a build but the master will no longer know about it.

slave.mark_as_idle() is where we check if the slave is in shutdown mode and 'kill' it if so, but from the slave object there's no way to remove it from the master's _all_slaves_by_url.  Still thinking about this.